src/cryptonote_protocol/block_queue.h

```diff
@@ -0,0 +1,386 @@
+// Copyright (c) 2017, The Monero Project
+// 
+// All rights reserved.
+// 
+// Redistribution and use in source and binary forms, with or without modification, are
+// permitted provided that the following conditions are met:
+// 
+// 1. Redistributions of source code must retain the above copyright notice, this list of
+//    conditions and the following disclaimer.
+// 
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list
+//    of conditions and the following disclaimer in the documentation and/or other
+//    materials provided with the distribution.
+// 
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+// 
+// Parts of this file are originally copyright (c) 2012-2013 The Cryptonote developers
+
+#pragma once
+
+#include <string>
+#include <ctime>
+#include <set>
+#include <unordered_map>
+#include <mutex>
+#include <boost/uuid/uuid.hpp>
+#include "cryptonote_protocol_defs.h"
+
+#undef MONERO_DEFAULT_LOG_CATEGORY
+#define MONERO_DEFAULT_LOG_CATEGORY "cn.block_queue"
+
+namespace std {
+  template<> struct hash<boost::uuids::uuid> {
+    std::size_t operator()(const boost::uuids::uuid &_v) const {
+      return reinterpret_cast<const std::size_t &>(_v);
+    }
+  };
+}
+namespace cryptonote
+{
+  class block_queue
+  {
+  public:
+    struct span
+    {
+      uint64_t start_block_height;
+      std::list<crypto::hash> hashes;
+      std::list<cryptonote::block_complete_entry> blocks;
```

Comment:
#### block_queue
My primary issue is that I may not understand the purpose of this class. It seems closer to a "cache" so that blocks can be received out-of-order, and then processed later after missing gaps are filled. If so -

The current patch is storing these chunks by first block height in a `std::set`, but then each entry in the set has an internal list of bocks that has to be inspected multiple times. Since `std::set` has an insertion hint, contiguous ranges of blocks can be inserted into the `std::set` with a total complexity logarithmic to the existing set size. Many of the linear operations become logarithmic too, and should have less code. Overlapping chunks can be easily handled too - ignore duplicates and insert into the gaps. Alternate blocks could also be handled, but might be a bit much.

The "penalty" is that the blocks have to be iterated during insertion, and each insertion requires an allocation. The increased size of the set is going to be minimal; 1million blocks is roughly 19 ops (which has lots of pointer traversals etc, but is still significantly smaller than scanning 1million ops in those other find functions).

#### Move Parameters
`&&` in this context means the parameter must be an rvalue. My perspective: by-value copies by default and the caller can optionally move the memory, whereas by-rvalue moves by default and the caller can optionally copy the memory. In the rvalue case, it is actually a compiler error if the caller does not explicitly make a copy:
```c++
void foo(std::string&&);

std::string mystring;
foo(mystring); // compiler error, mystring is not an rvalue
foo(std::string(mystring)); // copied
foo(std::move(mystring)); // moved
foo(std::string("a new string")); // temporary string is "moved"
foo("a different string"); // implicit constructor is called
```